cpp/ql/lib/semmle/code/cpp/models/implementations/Iterator.qll

```diff
@@ -123,14 +123,34 @@ class IteratorCrementNonMemberOperator extends Operator {
 }
 
 private class IteratorCrementNonMemberOperatorModel extends IteratorCrementNonMemberOperator,
-  DataFlowFunction
+  DataFlowFunction, SideEffectFunction, AliasFunction
 {
   override predicate hasDataFlow(FunctionInput input, FunctionOutput output) {
     input = getIteratorArgumentInput(this, 0) and
     output.isReturnValue()
     or
     input.isParameterDeref(0) and output.isReturnValueDeref()
   }
+
+  override predicate hasOnlySpecificReadSideEffects() { any() }
+
+  override predicate hasOnlySpecificWriteSideEffects() { any() }
+
+  override predicate hasSpecificReadSideEffect(ParameterIndex i, boolean buffer) {
+    i = 0 and buffer = false
+  }
+
+  override predicate hasSpecificWriteSideEffect(ParameterIndex i, boolean buffer, boolean mustWrite) {
+    // `buffer` must be `true` or `mustWrite` must be `false` to ensure that
+    // the IR alias analysis doesn't think that `it++` doesn't completely override
+    // the value of the iterator. Ideally, the IR alias analysis shouldn't deal with
+    // iterators, but this is necessary for taintflow to get any results in our iterator tests.
+    i = 0 and buffer = false and mustWrite = false
```

Comment:
From the comment I would have expected `i = 0 and (buffer = true or mustWrite = false)`?

Reply:
Oh, yeah I can see why you'd think that. What I meant was that we either need to model `operator++` on iterators as:
- An operation that overwrites a subpart of the iterator (that is, `buffer = true`) and it modifies it unconditionally (that is, `mustWrite = true`), or
- An operation that overwrites the entire iterator (that is, `buffer = false`), but it only does so conditionally (that is, `mustWrite = false`).

Both of those situations will cause the IR alias analysis to conclude that the previous value of what the iterator pointed to may still be live (which is what we want since we want a write to the iterator to be obtained even though we increment the iterator since some iterators also support `operator--`).

Since the output of `hasSpecificWriteSideEffect` decides if (and which kinds of) write side effects should be emitted for a call, having multiple results for this predicate will generate IR inconsistencies where the same argument generates multiple different write side effects.

Does that make sense?

Reply:
Yes that makes sense.

Reply:
Maybe try to slightly rephrase it 
```
We have two choices here: either `buffer` must be `true` or `mustWrite` must be `false` to ensure that the IR alias analysis doesn't think that `it++` doesn't completely override the value of the iterator. We choose `mustWrite` must be `false`.
```
Am I correct to assume that the value of `buffer` doesn't really matter in this case any more, but we still need to bind it and do so by picking `false`?

Reply:
I like that phrasing, yeah. I'll push that 👍

> Am I correct to assume that the value of `buffer` doesn't really matter in this case any more, but we still need to bind it and do so by picking `false`?

That's correct, yes. As long as `mustWrite` is `false` it doesn't actually matter what `buffer` is. That just decides whether we need to emit a `BufferReadSideEffect` instruction or an `IndirectReadSideEffect` instruction.

Reply:
^ I can add that explanation as well to the comment.

Reply:
Given what you did below, might as well replace this one too with
```
    // See the comment on `IteratorCrementMemberOperatorModel::hasSpecificWriteSideEffect`
    // for an explanation of these values.
```

Reply:
Fixed in https://github.com/github/codeql/pull/16175/commits/0ed07310243f6f8ebb66786915c5108b56724f73